File: ruby/day1.rb
def part1(input)
  sum = 0
  input.each_line {|line|
    matches = line.scan(/\d/)
    num = (matches.first + matches.last).to_i
    sum += num
  }
  puts "part 1: #{sum}"
end

def part2(input)
  sum = 0
  i = 0
  input.each_line {|line|
    i += 1
    line = line
      .gsub("one", "o1ne")
      .gsub("two", "t2wo")
      .gsub("three", "t3hree")
      .gsub("four", "f4our")
      .gsub("five", "f5ive")
      .gsub("six", "s6ix")
      .gsub("seven", "s7even")
      .gsub("eight", "e8ight")
      .gsub("nine", "n9ine")

    matches = line.scan(/\d/)
    num = (matches.first + matches.last).to_i
    sum += num
  }
  puts "part 2: #{sum}"
end

testInput = "1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"

part1(testInput)
part1(File.read("day1.txt"))

part2("two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen")
part2(File.read("day1.txt"))

File: zig/day1.zig
const std = @import("std");

fn part1(input: []const u8) u64 {
    var split = std.mem.splitScalar(u8, input, '\n');
    var sum: u64 = 0;
    while (split.next()) |line| {
        var first: ?u8 = null;
        var last: ?u8 = null;
        for (line) |chr| {
            if (std.ascii.isDigit(chr)) {
                if (first == null) {
                    first = chr;
                }
                last = chr;
            }
        }
        if (first == null or last == null) {
            continue;
        }
        sum += (first.? - 48) * 10 + (last.? - 48);
    }
    return sum;
}

pub fn main() !void {
    // const testInput =
    //     \\1abc2
    //     \\pqr3stu8vwx
    //     \\a1b2c3d4e5f
    //     \\treb7uchet
    // ;

    // std.debug.print("part1 test: {d}\n", .{part1(testInput)});

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const alloc = gpa.allocator();

    var realInput = try std.fs.cwd().readFileAlloc(alloc, "day1.txt", 100000);
    defer alloc.free(realInput);

    std.debug.print("part1: {d}\n", .{part1(realInput)});
}
